src/Lumper.Lib/BSP/BspFile.cs

```diff
@@ -35,45 +43,238 @@ public class BspFile
+                {
+                    // This should never happen since path depends on current
+                    // millisecond but whatever
+                    Logger.Error($"Backup path {backupPath} exists!");
+                    return;
+                }
+
+                try
+                {
+                    // TODO: go back to copy?
+                    if (FileStream is not null)
+                    {
+                        using FileStream backupStream = File.Open(backupPath, FileMode.Create);
+                        FileStream.Seek(0, SeekOrigin.Begin);
+                        FileStream.CopyTo(backupStream);
+                    }
+                    else
+                    {
+                        File.Copy(outPath, backupPath);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error(ex, "Failed to create backup file, not saving");
+                    return;
+                }
+
+                Logger.Info($"Created backup file {backupPath}");
+            }
+        }
+
+        Stream stream;
+        var overwritingOpenFile = false;
+        // Check if we're trying to write to the file we current have open.
+        // If so, saving will try to use data from that open filestream, so we
+        // can't write to it at the same time.
+        // Instead, write to a memory stream, then on success we'll close to
+        // open filestream, re-open it, and write the memory stream out to it.
+        if (File.Exists(outPath) && outPath == FilePath && FileStream is not null)
+        {
+            overwritingOpenFile = true;
+            stream = new MemoryStream();
+        }
+        else
+        {
+            // Overwrites file if exists
+            stream = File.Open(outPath, FileMode.Create);
+        }
+
+        var writer = new BspFileWriter(this, stream, handler, compress);
+        bool success;
+        try
+        {
+            success = writer.Save();
+        }
+        catch (Exception ex)
+        {
+            Logger.Error(ex, "Failed to write BSP file to stream");
+            success = false;
+        }
+
+        if (success)
+        {
+            if (overwritingOpenFile)
+            {
+                // Dispose of filestream we were just reading from
+                FileStream!.Dispose();
+
+                // Reopen, FileMode.Create overwrites
+                using FileStream fstream = File.Open(outPath, FileMode.Create);
+                stream.Seek(0, SeekOrigin.Begin);
+
+                // Copy memorystream we just wrote to over to new filestream, dispose
+                stream.CopyTo(fstream);
+                // Dispose of writer and underlying `stream`.
+                writer.Dispose();
+            }
+            else
+            {
+                // This is a filestream to a new file, just flush and close.
+                writer.Dispose();
+            }
```

Comment:
`using var writer...` instead of this?

Reply:
My next push changes how we do stream disposal quite a bit. I seem able to use `using writer` but the underlying stream has to be disposed of manually, since in some code-paths the stream is a writable FileStream that we want need to dispose of before opening a new stream to that same file.

Hopefully the save/load stuff is significantly easier to follow after this push!